Failed Clone on Windows (OSError: Invalid argument for path that contains asterisk)

When I clone on Windows, I am seeing the following error:

```
Cloning into 'C:\Users\pwintz\flashcards\ki\collection'...
Notes: 100%|##############################| 3607/3607 [00:02<00:00, 1686.82it/s]
HTML: 100%|#####################################| 11/11 [00:01<00:00,  7.55it/s]
Media: 100%|#############################| 3607/3607 [00:00<00:00, 23412.28it/s]
Decks:  12%|####3                               | 11/90 [00:00<00:03, 20.10it/s]2022-09-09 22:56:00.489 | WARNING  | ki:get_note_path:853 - Slug for 'QM6kTt.Nt*' is empty. Using guid as filename
Decks:  13%|####8                               | 12/90 [00:00<00:04, 17.77it/s]
Traceback (most recent call last):
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\Scripts\ki-script.py", line 33, in <module>
    sys.exit(load_entry_point('ki==0.0.4a0', 'console_scripts', 'ki')())
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\ki\__init__.py", line 1993, in clone
    raise err
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\ki\__init__.py", line 1979, in clone
    _, _ = _clone(
  File "<@beartype(ki._clone) at 0x22bd2204430>", line 104, in _clone
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\ki\__init__.py", line 2052, in _clone
    latent_links: Set[LatentSymlink] = write_repository(
  File "<@beartype(ki.write_repository) at 0x22bd21f9e50>", line 123, in write_repository
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\ki\__init__.py", line 1270, in write_repository
    latent_links: Set[LatentSymlink] = write_decks(
  File "<@beartype(ki.write_decks) at 0x22bd21f9f70>", line 121, in write_decks
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\ki\__init__.py", line 1379, in write_decks
    node_cids, node_notes, node_latent_links = write_deck_node_cards(
  File "<@beartype(ki.write_deck_node_cards) at 0x22bd21ff0d0>", line 158, in write_deck_node_cards
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\ki\__init__.py", line 1469, in write_deck_node_cards
    note_path: ExtantFile = F.write(note_path, payload)
  File "<@beartype(ki.functional.write) at 0x22bd21c3820>", line 52, in write
  File "C:\Users\pwintz\AppData\Roaming\Python\Python39\site-packages\ki\functional.py", line 164, in write
    with open(path, "w+", encoding="UTF-8") as f:
OSError: [Errno 22] Invalid argument: 'C:\\Users\\pwintz\\flashcards\\ki\\collection\\Life\\Countries\\QM6kTt.Nt*.md'

```
My guess is that the asterisk in the file name is causing problems.

I have reproduced this error in a test, and it does indeed fail reproducibly on Windows (not that I had any doubt, your bug reports are always impeccable).

```python
        """If there are e.g. asterisks in a `guid`, do we still write the note file successfully?"""
        ORIGINAL: SampleCollection = get_test_collection("original")
        col = Collection(ORIGINAL.col_file)
        timestamp_ns: int = time.time_ns()
        new_nids: Iterator[int] = itertools.count(int(timestamp_ns / 1e6))

        for nid, guid, mod, mid in col.db.execute("select id, guid, mod, mid from notes"):
            logger.debug(f"{nid = }")
            logger.debug(f"{mod = }")
            logger.debug(f"{mid = }")

        nid = next(new_nids)
        mod = next(new_nids)
        mid = 1645010146011
        guid = r"QM6kTt.Nt*"
        decknote = DeckNote(
            title="title",
            guid=guid,
            deck="Default",
            model="Basic",
            tags=[],
            fields={"Front": "", "Back": "b"},
        )
        note: Note = add_db_note(
            col,
            nid,
            guid,
            mid,
            mod=int(timestamp_ns // 1e9),
            usn=-1,
            tags=decknote.tags,
            fields=list(decknote.fields.values()),
            sfld="",
            csum=0,
            flags=0,
            data="",
        )
        notetype: Notetype = parse_notetype_dict(note.note_type())

        runner = CliRunner()
        with runner.isolated_filesystem(temp_dir=tmp_path):
            targetdir: ExtantDir = F.mkdir(F.test(Path("collection")))
            colnote = ColNote(
                n=note,
                new=True,
                deck="Default",
                title="None",
                markdown=False,
                notetype=notetype,
                sortf_text="",
            )
            note_path: NoFile = get_note_path(colnote, targetdir)
            payload: str = "payload"
>           note_path: ExtantFile = F.write(note_path, payload)

tests\test_ki.py:1798:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
<@beartype(ki.functional.write) at 0x223c22e0e50>:52: in write
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

path = NoFile('collection/QM6kTt.Nt*.md'), text = 'payload'

    @beartype
    def write(path: Union[ExtantFile, NoFile], text: str) -> ExtantFile:
        """Write text to a file."""
>       with open(path, "w+", encoding="UTF-8") as f:
E       OSError: [Errno 22] Invalid argument: 'collection\\QM6kTt.Nt*.md'

ki\functional.py:164: OSError
```

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -1,4 +1,4 @@
-pytest
+pytest==7.1.2
 pytest-mock
 pytest-timeout
 checksumdir==1.2.0

--- a/ki/__init__.py
+++ b/ki/__init__.py
@@ -26,6 +26,7 @@ import logging
 import secrets
 import sqlite3
 import hashlib
+import datetime
 import itertools
 import traceback
 import functools
@@ -152,7 +153,7 @@ REMOTE_NAME = "anki"
 BRANCH_NAME = "main"
 CHANGE_TYPES = "A D R M T".split()
 TQDM_NUM_COLS = 80
-MAX_FILENAME_LEN = 40
+MAX_FILENAME_LEN = 60
 IGNORE_DIRECTORIES = set([GIT, KI, MEDIA])
 IGNORE_FILES = set([GITIGNORE_FILE, GITMODULES_FILE, MODELS_FILE])
 HEAD_SUFFIX = Path("ki-head")
@@ -828,7 +829,7 @@ def validate_decknote_fields(notetype: Notetype, decknote: DeckNote) -> List[War
 # first field to get a unique filename, if they exist. Note that we must still
 # treat the case where all fields are images!
 @beartype
-def get_note_path(colnote: ColNote, deck_dir: ExtantDir, card_name: str = "") -> NoFile:
+def get_note_path(colnote: ColNote, payload: str, deck_dir: ExtantDir, card_name: str = "") -> NoFile:
     """Get note path from sort field text."""
     field_text = colnote.sortf_text
 
@@ -844,13 +845,26 @@ def get_note_path(colnote: ColNote, deck_dir: ExtantDir, card_name: str = "") ->
     name = field_text[:MAX_FILENAME_LEN]
     slug = F.slugify(name)
 
-    # Make it so `slug` cannot possibly be an empty string, because then we get
-    # a `Path('.')` which is a bug, and causes a runtime exception.  If all
-    # else fails, generate a random hex string to use as the filename.
+    # If the slug is still empty, use all the fields.
     if len(slug) == 0:
-        slug = str(colnote.n.guid)
-        msg = f"Slug for '{colnote.n.guid}' is empty. Using guid as filename"
-        logger.warning(msg)
+        contents = " ".join(colnote.n.values())
+        name = contents[:MAX_FILENAME_LEN]
+        slug = F.slugify(name)
+
+    # Make it so `slug` cannot possibly be an empty string, because then we get
+    # a `Path('.')` which is a bug, and causes a runtime exception. If all else
+    # fails, use the notetype name, hash of the payload, and creation date.
+    if len(slug) == 0:
+        blake2 = hashlib.blake2s()
+        blake2.update(colnote.n.guid.encode("UTF-8"))
+        slug: str = f"{colnote.notetype.name}--{blake2.hexdigest()}"
+
+        # Note IDs are in milliseconds.
+        dt = datetime.datetime.fromtimestamp(colnote.n.id / 1000.0)
+        slug += "--" + dt.strftime("%Y-%m-%d--%Hh-%Mm-%Ss")
+
+        logger.warning(f"Slug for note with guid '{colnote.n.guid}' is empty...")
+        logger.warning(f"Using blake2 hash of guid as filename: '{slug}'")
 
     if card_name != "":
         slug = f"{slug}_{card_name}"
@@ -867,7 +881,7 @@ def get_note_path(colnote: ColNote, deck_dir: ExtantDir, card_name: str = "") ->
 
 
 @beartype
-def backup(kirepo: KiRepo) -> None:
+def backup(kirepo: KiRepo) -> int:
     """Backup collection to `.ki/backups`."""
     md5sum = F.md5(kirepo.col_file)
     name = f"{md5sum}.anki2"
@@ -880,10 +894,11 @@ def backup(kirepo: KiRepo) -> None:
     # location.
     if isinstance(backup_file, ExtantFile):
         echo("Backup already exists.")
-        return
+        return 1
 
     echo(f"Writing backup of .anki2 file to '{backup_file}'")
     F.copyfile(kirepo.col_file, kirepo.backups_dir, name)
+    return 0
 
 
 @beartype
@@ -1464,8 +1479,8 @@ def write_deck_node_cards(
         colnote: ColNote = colnotes[card.nid]
 
         if card.nid not in written_notes:
-            note_path: NoFile = get_note_path(colnote, deck_dir)
             payload: str = get_note_payload(colnote, tidy_field_files)
+            note_path: NoFile = get_note_path(colnote, payload, deck_dir)
             note_path: ExtantFile = F.write(note_path, payload)
             written_notes[card.nid] = WrittenNoteFile(did, note_path)
         else:
@@ -1483,7 +1498,8 @@ def write_deck_node_cards(
             template: TemplateDict = card.template()
             name: str = template["name"]
 
-            note_path: NoFile = get_note_path(colnote, deck_dir, name)
+            payload: str = get_note_payload(colnote, tidy_field_files)
+            note_path: NoFile = get_note_path(colnote, payload, deck_dir, name)
             abs_target: ExtantFile = written_notes[card.nid].file
             distance = len(note_path.parent.relative_to(targetdir).parts)
             up_path = Path("../" * distance)

--- a/tests/test_integration.py
+++ b/tests/test_integration.py
@@ -1497,7 +1497,6 @@ def test_push_writes_media(tmp_path: Path):
     with runner.isolated_filesystem(temp_dir=tmp_path):
 
         # Clone.
-        logger.debug(f"First clone...")
         clone(runner, MEDIACOL.col_file)
 
         # Add a new note file containing media, and the corresponding media file.
@@ -1505,13 +1504,11 @@ def test_push_writes_media(tmp_path: Path):
         media_note_path = root / MEDIACOL.repodir / "Default" / MEDIA_NOTE
         media_file_path = root / MEDIACOL.repodir / "Default" / MEDIA / MEDIA_FILENAME
         onesec_file = root / MEDIACOL.repodir / "Default" / MEDIA / "1sec.mp3"
-        logger.debug(f"Copying media file to '{media_file_path}'")
         shutil.copyfile(MEDIA_NOTE_PATH, media_note_path)
         shutil.copyfile(MEDIA_FILE_PATH, media_file_path)
         os.chdir(MEDIACOL.repodir)
 
         mode: int = ki.filemode(onesec_file)
-        logger.warning(f"1sec.mp3: {mode = }")
 
         # Commit the additions.
         repo = git.Repo(F.cwd())
@@ -1520,7 +1517,6 @@ def test_push_writes_media(tmp_path: Path):
         repo.close()
 
         mode: int = ki.filemode(onesec_file)
-        logger.warning(f"1sec.mp3: {mode = }")
 
         # Push the commit.
         out = push(runner)
@@ -1530,7 +1526,6 @@ def test_push_writes_media(tmp_path: Path):
         F.rmtree(F.test(Path(MEDIACOL.repodir)))
 
         # Re-clone the pushed collection.
-        logger.debug(f"Second clone...")
         out = clone(runner, MEDIACOL.col_file)
 
         # Check that added note and media file exist.
@@ -1865,9 +1860,9 @@ def test_push_correctly_encodes_quotes_in_html_tags(tmp_path: Path):
         # Clone collection in cwd.
         clone(runner, BROKEN.col_file)
         os.chdir(BROKEN.repodir)
-
         note_file = (
-            Path("🧙‍Recommendersysteme") / "wie-sieht-die-linkstruktur-von-einem-hub.md"
+            Path("🧙‍Recommendersysteme")
+            / "wie-sieht-die-linkstruktur-von-einem-hub-in-einem-web-graphe.md"
         )
         with open(note_file, "r", encoding="UTF-8") as read_f:
             contents = read_f.read().replace("guter", "guuter")

--- a/tests/test_ki.py
+++ b/tests/test_ki.py
@@ -21,10 +21,12 @@ from lark.exceptions import UnexpectedToken
 from loguru import logger
 from pytest_mock import MockerFixture
 from click.testing import CliRunner
-from anki.collection import Collection
+
+import anki.collection
+from anki.collection import Collection, Note
 
 from beartype import beartype
-from beartype.typing import List, Union, Set, Iterator
+from beartype.typing import List, Union, Set, Iterator, Tuple
 
 import ki
 import ki.maybes as M
@@ -78,6 +80,7 @@ from ki import (
     copy_media_files,
     diff2,
     _clone,
+    add_db_note,
 )
 from ki.types import (
     NoFile,
@@ -706,14 +709,14 @@ def test_slugify_handles_html_tags():
 
 
 @beartype
-def get_colnote_with_sortf_text(sortf_text: str) -> ColNote:
+def get_colnote_with_sortf_text(sortf_text: str) -> Tuple[Collection, ColNote]:
     ORIGINAL: SampleCollection = get_test_collection("original")
     col = open_collection(ORIGINAL.col_file)
     note = col.get_note(set(col.find_notes("")).pop())
 
     notetype: Notetype = parse_notetype_dict(note.note_type())
     deck = col.decks.name(note.cards()[0].did)
-    return ColNote(
+    return col, ColNote(
         n=note,
         new=False,
         deck=deck,
@@ -724,22 +727,50 @@ def get_colnote_with_sortf_text(sortf_text: str) -> ColNote:
     )
 
 
+@beartype
+def get_basic_colnote_with_fields(front: str, back: str) -> Tuple[Collection, ColNote]:
+    ORIGINAL: SampleCollection = get_test_collection("original")
+    col = open_collection(ORIGINAL.col_file)
+    note = col.get_note(set(col.find_notes("")).pop())
+    note["Front"] = front
+    note["Back"] = back
+    note.flush()
+
+    notetype: Notetype = parse_notetype_dict(note.note_type())
+    deck = col.decks.name(note.cards()[0].did)
+    return col, ColNote(
+        n=note,
+        new=False,
+        deck=deck,
+        title="",
+        markdown=False,
+        notetype=notetype,
+        sortf_text=note[notetype.sortf.name],
+    )
+
+
 def test_get_note_path_produces_nonempty_filenames():
     field_text = '<img src="card11front.jpg" />'
-    colnote = get_colnote_with_sortf_text(field_text)
+    col, colnote = get_colnote_with_sortf_text(field_text)
 
     runner = CliRunner()
     with runner.isolated_filesystem():
         deck_dir: ExtantDir = F.force_mkdir(Path("a"))
 
-        path: ExtantFile = get_note_path(colnote, deck_dir)
+        payload = get_note_payload(colnote, {})
+        path: ExtantFile = get_note_path(colnote, payload, deck_dir)
         assert path.name == "img-srccard11frontjpg.md"
 
         # Check that it even works if the field is empty.
-        empty_colnote = get_colnote_with_sortf_text("")
-        path: ExtantFile = get_note_path(empty_colnote, deck_dir)
+        col, empty_colnote = get_colnote_with_sortf_text("")
+        payload = get_note_payload(empty_colnote, {})
+        path: ExtantFile = get_note_path(empty_colnote, payload, deck_dir)
         assert ".md" in str(path)
         assert f"{os.sep}a{os.sep}" in str(path)
+
+        col, empty = get_basic_colnote_with_fields("", "")
+        payload = get_note_payload(empty, {})
+        path: ExtantFile = get_note_path(empty, payload, deck_dir)
 
 
 @beartype
@@ -792,7 +823,7 @@ def get_diff2_args() -> DiffReposArgs:
     return DiffReposArgs(remote_repo, parser, transformer)
 
 
-def test_diff2_shows_no_changes_when_no_changes_have_been_made(capfd, tmp_path):
+def test_diff2_shows_no_changes_when_no_changes_have_been_made(tmp_path: Path):
     ORIGINAL = get_test_collection("original")
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -807,16 +838,12 @@ def test_diff2_shows_no_changes_when_no_changes_have_been_made(capfd, tmp_path):
             args.parser,
             args.transformer,
         )
-        del args
-        gc.collect()
 
         changed = [str(delta.path) for delta in deltas]
-        captured = capfd.readouterr()
         assert changed == []
-        assert "last_push" not in captured.err
 
 
-def test_diff2_yields_a_warning_when_a_file_cannot_be_found(tmp_path):
+def test_diff2_yields_a_warning_when_a_file_cannot_be_found(tmp_path: Path):
     ORIGINAL = get_test_collection("original")
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -847,7 +874,7 @@ def test_diff2_yields_a_warning_when_a_file_cannot_be_found(tmp_path):
         assert "note123412341234.md" in str(warning)
 
 
-def test_unsubmodule_repo_removes_gitmodules(tmp_path):
+def test_unsubmodule_repo_removes_gitmodules(tmp_path: Path):
     """
     When you have a ki repo with submodules, does calling
     `unsubmodule_repo()`on it remove them? We test this by checking if the
@@ -917,18 +944,41 @@ def test_diff2_handles_submodules():
             assert delta.path.is_file()
 
 
-def test_backup_is_no_op_when_backup_already_exists(capfd):
-    """Do we print a nice message when we backup an already-backed-up file?"""
+def test_backup_is_no_op_when_backup_already_exists(mocker: MockerFixture):
+    """
+    Does the second subsequent backup call register as redundant (we assume a
+    nice error message is displayed if so)?
+    """
     ORIGINAL = get_test_collection("original")
     runner = CliRunner()
     with runner.isolated_filesystem():
         clone(runner, ORIGINAL.col_file)
         os.chdir(ORIGINAL.repodir)
         kirepo: KiRepo = M.kirepo(F.cwd())
-        backup(kirepo)
-        backup(kirepo)
-        captured = capfd.readouterr()
-        assert "Backup already exists." in captured.out
+
+        mock = mocker.patch("ki.echo")
+
+        returncode: int = backup(kirepo)
+        assert returncode == 0
+        calls = list(mock.call_args_list)
+        assert len(calls) == 1
+        call = calls[0]
+        args = call.args
+        assert len(call.args) == 1
+        displayed = call.args[0]
+        assert isinstance(displayed, str)
+        assert "Writing backup" in displayed
+
+        returncode = backup(kirepo)
+        assert returncode == 1
+        calls = list(mock.call_args_list)
+        assert len(calls) == 2
+        call = calls[-1]
+        args = call.args
+        assert len(call.args) == 1
+        displayed = call.args[0]
+        assert isinstance(displayed, str)
+        assert "Backup already exists" in displayed
 
 
 def test_git_pull():
@@ -960,12 +1010,13 @@ def test_git_pull():
 def test_get_note_path():
     """Do we add ordinals to generated filenames if there are duplicates?"""
     runner = CliRunner()
-    colnote = get_colnote_with_sortf_text("a")
+    col, colnote = get_colnote_with_sortf_text("a")
     with runner.isolated_filesystem():
         deck_dir = F.cwd()
         dupe_path = deck_dir / "a.md"
         dupe_path.write_text("ay")
-        note_path = get_note_path(colnote, deck_dir)
+        payload = get_note_payload(colnote, {})
+        note_path = get_note_path(colnote, payload, deck_dir)
         assert str(note_path.name) == "a_1.md"
 
 
@@ -1126,7 +1177,7 @@ def test_write_repository_propogates_errors_from_get_colnote(mocker: MockerFixtu
         assert "'bad_field_key'" in str(error.exconly())
 
 
-def test_maybe_kirepo_displays_nice_errors(tmp_path):
+def test_maybe_kirepo_displays_nice_errors(tmp_path: Path):
     """Does a nice error get printed when kirepo metadata is missing?"""
     ORIGINAL = get_test_collection("original")
     runner = CliRunner()
@@ -1170,7 +1221,6 @@ def test_maybe_kirepo_displays_nice_errors(tmp_path):
             M.kirepo(targetdir)
         assert "A directory was expected" in str(error.exconly())
         assert "'.ki/backups'" in str(error.exconly())
-        logger.debug(Path.cwd())
         gc.collect()
         F.rmtree(targetdir)
 
@@ -1228,7 +1278,7 @@ def test_maybe_kirepo_displays_nice_errors(tmp_path):
         F.rmtree(targetdir)
 
 
-def test_get_target(tmp_path):
+def test_get_target(tmp_path: Path):
     """Do we print a nice error when the targetdir is nonempty?"""
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1241,7 +1291,7 @@ def test_get_target(tmp_path):
         assert "file" in str(error.exconly())
 
 
-def test_maybe_emptydir(tmp_path):
+def test_maybe_emptydir(tmp_path: Path):
     """Do we print a nice error when the directory is unexpectedly nonempty?"""
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1252,7 +1302,7 @@ def test_maybe_emptydir(tmp_path):
         assert str(Path.cwd()) in str(error.exconly()).replace("\n", "")
 
 
-def test_maybe_emptydir_handles_non_directories(tmp_path):
+def test_maybe_emptydir_handles_non_directories(tmp_path: Path):
     """Do we print a nice error when the path is not a directory?"""
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1266,7 +1316,7 @@ def test_maybe_emptydir_handles_non_directories(tmp_path):
 @pytest.mark.skipif(
     sys.platform == "win32", reason="Windows does not have `os.mkfifo()`."
 )
-def test_maybe_xdir(tmp_path):
+def test_maybe_xdir(tmp_path: Path):
     """Do we print a nice error when there is a non-file non-directory thing?"""
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1280,7 +1330,7 @@ def test_maybe_xdir(tmp_path):
 @pytest.mark.skipif(
     sys.platform == "win32", reason="Windows does not have `os.mkfifo()`."
 )
-def test_maybe_xfile(tmp_path):
+def test_maybe_xfile(tmp_path: Path):
     """Do we print a nice error when there is a non-file non-directory thing?"""
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1410,7 +1460,7 @@ def test_get_colnote_propagates_errors_key_errors_from_sort_field(
     assert "'bad_field_key'" in str(error.exconly())
 
 
-def test_nopath(tmp_path):
+def test_nopath(tmp_path: Path):
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
         file = Path("file")
@@ -1419,7 +1469,7 @@ def test_nopath(tmp_path):
             M.nopath(file)
 
 
-def test_copy_kirepo(tmp_path):
+def test_copy_kirepo(tmp_path: Path):
     """
     Do errors in `M.nopath()` call in `copy_kirepo()` get forwarded to
     the caller and printed nicely?
@@ -1449,7 +1499,7 @@ def test_copy_kirepo(tmp_path):
         assert ".ki" in str(error.exconly())
 
 
-def test_filter_note_path(tmp_path):
+def test_filter_note_path(tmp_path: Path):
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
         directory = Path("directory")
@@ -1466,7 +1516,7 @@ def test_filter_note_path(tmp_path):
         assert str(Path("directory") / "file") in str(warning)
 
 
-def test_get_models_recursively(tmp_path):
+def test_get_models_recursively(tmp_path: Path):
     ORIGINAL = get_test_collection("original")
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1484,7 +1534,7 @@ def test_get_models_recursively(tmp_path):
 
 
 def test_get_models_recursively_prints_a_nice_error_when_models_dont_have_a_name(
-    tmp_path,
+    tmp_path: Path,
 ):
     ORIGINAL = get_test_collection("original")
     runner = CliRunner()
@@ -1501,7 +1551,7 @@ def test_get_models_recursively_prints_a_nice_error_when_models_dont_have_a_name
         assert "1645010146011" in str(error.exconly())
 
 
-def test_copy_repo_handles_submodules(tmp_path):
+def test_copy_repo_handles_submodules(tmp_path: Path):
     ORIGINAL = get_test_collection("original")
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1531,7 +1581,7 @@ def test_copy_repo_handles_submodules(tmp_path):
 @pytest.mark.skipif(
     sys.platform == "win32", reason="Windows does not have `os.mkfifo()`."
 )
-def test_ftest_handles_strange_paths(tmp_path):
+def test_ftest_handles_strange_paths(tmp_path: Path):
     """Do we print a nice error when there is a non-file non-directory thing?"""
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
@@ -1547,7 +1597,7 @@ def test_fparent_handles_fs_root():
     assert str(parent) == root
 
 
-def test_fmkleaves_handles_collisions(tmp_path):
+def test_fmkleaves_handles_collisions(tmp_path: Path):
     runner = CliRunner()
     with runner.isolated_filesystem(temp_dir=tmp_path):
         root = F.test(F.cwd())
@@ -1584,7 +1634,7 @@ def test_copy_media_files_returns_nice_errors():
         assert "bad Anki collection media directory" in str(error.exconly())
 
 
-def test_write_repository_displays_missing_media_warnings(capfd):
+def test_write_repository_displays_missing_media_warnings(mocker: MockerFixture):
     MEDIACOL: SampleCollection = get_test_collection("media")
     runner = CliRunner()
     with runner.isolated_filesystem():
@@ -1605,14 +1655,17 @@ def test_write_repository_displays_missing_media_warnings(capfd):
             if path.is_file():
                 os.remove(path)
 
+        mock = mocker.patch("click.secho")
         write_repository(
             MEDIACOL.col_file, targetdir, leaves, media_dir, silent=False, verbose=True
         )
+        calls = list(mock.call_args_list)
+        args: List[str] = list(map(lambda call: str(call.args), calls))
+        displayed = "\n".join(args)
 
-        missing = Path(MEDIACOL.media_directory_name) / "1sec.mp3"
-        captured = capfd.readouterr()
-        assert "Missing or bad media file" in captured.out
-        assert str(missing) in captured.out
+        assert "Missing or bad media file" in displayed
+        assert MEDIACOL.media_directory_name in displayed
+        assert "1sec.mp3" in displayed
 
 
 def test_copy_media_files_finds_notetype_media():
@@ -1663,7 +1716,6 @@ def test_copy_repo_preserves_git_symlink_file_modes(tmp_path: Path):
         # Check that filemode is initially 120000.
         repo = git.Repo(MEDIACOL.repodir)
         onesec_file = F.working_dir(repo) / "Default" / MEDIA / "1sec.mp3"
-        logger.debug(f"{onesec_file = }")
         mode = ki.filemode(onesec_file)
         assert mode == 120000
 
@@ -1671,6 +1723,59 @@ def test_copy_repo_preserves_git_symlink_file_modes(tmp_path: Path):
         ref = M.head_repo_ref(repo)
         ephem = ki.copy_repo(ref, "filemode-test")
         onesec_file = F.working_dir(ephem) / "Default" / MEDIA / "1sec.mp3"
-        logger.debug(f"{onesec_file = }")
         mode = ki.filemode(onesec_file)
         assert mode == 120000
+
+
+def test_write_deck_node_cards_does_not_fail_due_to_special_characters_in_paths_on_windows(
+    tmp_path: Path,
+):
+    """If there are e.g. asterisks in a `guid`, do we still write the note file successfully?"""
+    ORIGINAL: SampleCollection = get_test_collection("original")
+    col = Collection(ORIGINAL.col_file)
+    timestamp_ns: int = time.time_ns()
+    new_nids: Iterator[int] = itertools.count(int(timestamp_ns / 1e6))
+
+    nid = next(new_nids)
+    mod = next(new_nids)
+    mid = 1645010146011
+    guid = r"QM6kTt.Nt*"
+    decknote = DeckNote(
+        title="title",
+        guid=guid,
+        deck="Default",
+        model="Basic",
+        tags=[],
+        fields={"Front": "", "Back": "b"},
+    )
+    note: Note = add_db_note(
+        col,
+        nid,
+        guid,
+        mid,
+        mod=int(timestamp_ns // 1e9),
+        usn=-1,
+        tags=decknote.tags,
+        fields=list(decknote.fields.values()),
+        sfld="",
+        csum=0,
+        flags=0,
+        data="",
+    )
+    notetype: Notetype = parse_notetype_dict(note.note_type())
+
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        targetdir: ExtantDir = F.mkdir(F.test(Path("collection")))
+        colnote = ColNote(
+            n=note,
+            new=True,
+            deck="Default",
+            title="None",
+            markdown=False,
+            notetype=notetype,
+            sortf_text="",
+        )
+        payload: str = "payload"
+        note_path: NoFile = get_note_path(colnote, payload, targetdir)
+        note_path: ExtantFile = F.write(note_path, payload)

